### PR TITLE
Support ip address in server_name with service-identity

### PR DESCRIPTION
**Pull request by GalaxySnail**
Depends on #389. Once #389 is merged, I will rebase this PR.

### CODE REVIEW

File: tests/test_tls.py [hunk 1]
```diff
@@ -1420,6 +1402,32 @@ def test_verify_subject_with_subjaltname(self):
...
+            common_name="1.2.3.4",
+        )
+        cadata = certificate.public_bytes(serialization.Encoding.PEM)
+
+        with patch("aioquic.tls.utcnow") as mock_utcnow:
+            mock_utcnow.return_value = certificate.not_valid_before
+
+            # valid
+            verify_certificate(
+                cadata=cadata, certificate=certificate, server_name="1.2.3.4"
```
Comment (jlaine) [inline tests/test_tls.py:1421]: As explained, `server_name` is not meant to hold an IP address:
- search through tls.py and you'll see it's being used for SNI.
- look at asyncio/client.py and you'll see the `QuicConfiguration.server_name` is set to `None` if the host is an IP address

**Comment by jlaine** [comment]
Do you plan to complete this PR?

**Comment by GalaxySnail** [comment]
I don't know how to deal with them:

> `server_name` is (currently) not meant to hold an IP address:
> 
> * search through tls.py and you'll see it's being used for SNI.
> * look at asyncio/client.py and you'll see the `QuicConfiguration.server_name` is set to `None` if the host is an IP address

Please feel free to close it, I can submit a new PR if I'm ready.

**Comment by jlaine** [comment]
Could you at the very least let me know of a public server which uses an IP Address in its SAN?

**Comment by GalaxySnail** [comment]
AFAIK cloudflare's https://1.1.1.1 is a good example.

**Comment by jlaine** [comment]
> AFAIK cloudflare's https://1.1.1.1 is a good example.

Thanks, that's very useful!

I have changed the way `QuicConfiguration.server_name` is used, so that it can now be set to:

- A DNS name, in which case it is used bother for Server Name Indication and to validate the server's certificate.
- An IP address, in which case it is used to validate the server's certificate.
- `None`.

The `asyncio` API will now always populate `QuicConfiguration.server_name` with the host name, regardless of whether it is a DNS name or an IP address.

**Pull request merged**